### ISSUE TITLE
refactor: investigate possibility of reusing `//components/autofill` instead of maintaining a copy

### ISSUE THREAD

**Issue by ckerr** [electron/electron#31697]
suggested by @deepak1556 in https://github.com/electron/electron/pull/31676#pullrequestreview-796005562.

**Comment by ckerr** [comment]
See also https://github.com/electron/electron/blob/main/patches/chromium/put_back_deleted_colors_for_autofill.patch which comments:

> https://chromium-review.googlesource.com/c/chromium/src/+/1652925 removed colors as they are no longer needed in chromium but our autofill implementation uses them.  This patch can be removed if we refactor our autofill implementation to work like Chromium's

**Comment by jkleinsc** [comment]
This is still something we should do.

**Comment by deepak1556** [comment]
Relevant

**Comment by codebytere** [comment]
dug into this. what we call autofill is only the `<datalist>` dropdown - a renderer agent, a browser-side driver + factory, and the popup (controller + views), ~1750 lines in all.

~1100 of that is the popup, and `//components/autofill` has nothing to replace it with. Chrome's popup controller and views live in [`//chrome/browser/ui/autofill`](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/autofill/autofill_popup_controller_impl.cc) and [`//chrome/browser/ui/views/autofill/popup`](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/views/autofill/popup/) and pull in `Profile`, `BrowserView`, `ThemeService` and friends, and ours also paints into OSR, which upstream doesn't do at all. so that part stays regardless.

the agent and driver (~650 lines) could be swapped for upstream's the way Android WebView does it: upstream [`AutofillAgent` with null password agents](https://source.chromium.org/chromium/chromium/src/+/main:android_webview/renderer/aw_render_frame_ext.cc) in the renderer, and a mostly-stubbed [`ContentAutofillClient`](https://source.chromium.org/chromium/chromium/src/+/main:components/android_autofill/browser/android_autofill_client.cc) plus a small `AutofillManager` subclass in the browser that reads `FormFieldData::datalist_options()`. i don't think that's a good trade for us though:

- [`AutofillClient`](https://source.chromium.org/chromium/chromium/src/+/main:components/autofill/core/browser/foundations/autofill_client.h) has 43 pure virtuals and ~200 commits to the header in the last two years, [`AutofillManager`](https://source.chromium.org/chromium/chromium/src/+/main:components/autofill/core/browser/foundations/autofill_manager.h) 27 `*Impl` hooks and ~100 commits. WebView's stub client + manager took ~115 commits over the same window just to keep compiling. for comparison our agent/driver took +244/-116 lines across every roll since this was opened.
- upstream's agent extracts forms whenever a page loads or its forms change, and the base `AutofillManager` parses them with the field heuristics, in every renderer, with no knob to turn that off. ours does nothing until someone interacts with a `<datalist>` field.
- `AutofillManager` also [queries the crowdsourcing server](https://source.chromium.org/search?q=StartQueryRequest%20f:foundations%2Fautofill_manager.cc) while parsing unless `kAutofillServerCommunication` is disabled, so we'd carry that override forever.
- it's roughly LOC-neutral (WebView's stubs are ~650 lines) and doesn't remove any patches.

the one concrete cost mentioned above, `port_autofill_colors_to_the_color_pipeline.patch`, can go though - Chrome's popup just uses `ui::kColorDropdown*` now, so pointing ours at those ids drops the patch. i'll put that up separately. imo that's enough to close this out once it lands; should that change, we can revisit.
